0025-nat-traversal.adoc

```diff
@@ -0,0 +1,161 @@
+Bob
+
+STUN cannot be used to traverse NAT configurations that randomnly assign port and/or IP for new connections.
+A symmetric NAT will randomnly generate a new port and/or public IP for each outbound connection made by Alice.
+This prevents Bob from determining the external IP and ports assigned to her by her NAT.
+Symmetric NAT is commonly used by mobile network providers.
+
+- + Does not require the user to configure NAT settiThere is no single perfect solution to solving the NAT traversal problem.ngs
+
+- + Alice and Bob can directly communicate through a direct connection
+
+- - Requires a third peer to facilitate a connection between two peers
+
+- - Cannot traverse symmetric, cone and restricted cone type NATs
+
+- - Requires a third peer to start a connection between Alice and Bob
+
+- - Requires the third peer to know its inbound ports (either through another STUN/TURN or port forwarding)
+
+=== TURN
+
+If STUN cannot be used to traverse a NAT, a relay peer can be used a relay between the comit_nodes.
+When Bob isn’t able to listen on a public address, Bob can dial out to a relay peer, which will keep a long-lived connection open.
+Other peers will be able to dial through the relay peer, which will forward traffic to Bob.
+
+- + Can be used to traverse all commonly know NAT configurations
+
+- + Does not require the use to have configure NAT settings
+
+- - Requires a third peer to serve as a relay for communications between Alice and Bob
+
+- - Requires the third peer to know its inbound ports (either through another STUN/TURN or port forwarding)
+
+
+== Recommendation
+
+Manual and automatic port forwarding solve the NAT traversal problem through hardware configuration.
+We cannot rely on all users to have access to their network hardware and therefore we will need to implement alternative NAT traversal techniques.
+
+Reverse connections cannot be used because many of our protocols require both peers to be able to initate a connection with each other.
+This means that COMIT protocols will have to redesigned to use the connection initiated by dialer to complete a SWAP.
+Peers behind a NAT device will be unable to play an active role in the COMIT network.
+
+In order to allow comit nodes to traverse any NAT configuration the TURN protocol will need to be implemented.
+ICE recommends that peers attempt to connect using STUN if possible only use TURN a connection using STUN cannot be established.
+STUN is preferred over TURN for performance and efficiency reasons as it does require a third party to relay communications between nodes.
+
+Although TURN servers can resolve the NAT traversal problem, at least one STUN or TURN server that knows its own external ports and IP will be required to bootstrap the network and to restart the network in case it goes down.
+
+Users who have knowledge of their public IP and ports should have the option to run as STUN and/or TURN servers.
```

Comment:
I'm not sure if I follow :) 

Do you propose that we provide both solutions (STUN and TURN)? 

Can you rephrase this section so that it is clear what your recommendation is please? Same goes for the next section.

Reply:
yes we should ideally provide both solutions if want to be able to traverse most known NAT configurations. I'll reword this section to make it clear